raw-spans-grouper/src/test/java/org/hypertrace/core/rawspansgrouper/RawSpansGrouperTest.java

```diff
@@ -0,0 +1,118 @@
+  public void whenRawSpansAreReceivedWithInactivityExpectTraceToBeOutput(@TempDir Path tempDir) {
+    File file = tempDir.resolve("state").toFile();
+
+    RawSpansGrouper underTest = new RawSpansGrouper(ConfigClientFactory.getClient());
+    Config config = ConfigFactory.parseURL(
+        getClass().getClassLoader().getResource("configs/raw-spans-grouper/application.conf"));
+
+    Map<String, Object> baseProps = underTest.getBaseStreamsConfig();
+    Map<String, Object> streamsProps = underTest.getStreamsConfig(config);
+    baseProps.forEach(streamsProps::put);
+    Map<String, Object> mergedProps = streamsProps;
+
+    mergedProps.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, SpecificAvroSerde.class);
+    mergedProps.put(RawSpanGrouperConstants.RAW_SPANS_GROUPER_JOB_CONFIG, config);
+    mergedProps.put(StreamsConfig.STATE_DIR_CONFIG, file.getAbsolutePath());
+
+    StreamsBuilder streamsBuilder = underTest
+        .buildTopology(mergedProps, new StreamsBuilder(), new HashMap<>());
+
+    Properties props = new Properties();
+    mergedProps.forEach(props::put);
+
+    Serde defaultValueSerde = new StreamsConfig(mergedProps).defaultValueSerde();
+
+    TopologyTestDriver td = new TopologyTestDriver(streamsBuilder.build(), props);
+    TestInputTopic<String, RawSpan> inputTopic = td
+        .createInputTopic(config.getString(RawSpanGrouperConstants.INPUT_TOPIC_CONFIG_KEY),
+            Serdes.String().serializer(), defaultValueSerde.serializer());
+
+    TestOutputTopic outputTopic = td
+        .createOutputTopic(config.getString(RawSpanGrouperConstants.OUTPUT_TOPIC_CONFIG_KEY),
+            Serdes.String().deserializer(), defaultValueSerde.deserializer());
+
+    RawSpan span1 = RawSpan.newBuilder().setTraceId(ByteBuffer.wrap("trace-1".getBytes()))
+        .setCustomerId("customer1").setEvent(createEvent("event-1", "customer1")).build();
+    RawSpan span2 = RawSpan.newBuilder().setTraceId(ByteBuffer.wrap("trace-1".getBytes()))
+        .setCustomerId("customer1").setEvent(createEvent("event-2", "customer1")).build();
+    RawSpan span3 = RawSpan.newBuilder().setTraceId(ByteBuffer.wrap("trace-1".getBytes()))
+        .setCustomerId("customer1").setEvent(createEvent("event-3", "customer1")).build();
+    RawSpan span4 = RawSpan.newBuilder().setTraceId(ByteBuffer.wrap("trace-2".getBytes()))
+        .setCustomerId("customer1").setEvent(createEvent("event-4", "customer1")).build();
+    RawSpan span5 = RawSpan.newBuilder().setTraceId(ByteBuffer.wrap("trace-2".getBytes()))
+        .setCustomerId("customer1").setEvent(createEvent("event-5", "customer1")).build();
+
+    inputTopic.pipeInput("trace-1", span1);
+    inputTopic.pipeInput("trace-2", span4);
+    td.advanceWallClockTime(Duration.ofSeconds(1));
+    inputTopic.pipeInput("trace-1", span2);
+    // the next advance should trigger a punctuate call and emit a trace with 2 spans
+    td.advanceWallClockTime(Duration.ofSeconds(32));
```

Comment:
Do you think, we should cover `Duration.ofSeconds(25) - less than 30 (grouping window time)`, so puncuator kicks `25 + 30`?

Reply:
added it. The timing aspects make the tests a little flaky